File: Java/src/Minimum Window Substring.java
import java.util.ArrayList;
import java.util.HashMap;

public class Solution
{
  public String minWindow(String S, String T)
  {
    int start = 0;
    int end = S.length();
    int max = 0;
    int min = S.length();
    Map<Character, List<Integer>> dp = new HashMap<>();
    List<Character> listT = new ArrayList<>();

    for (int i = 0; i < T.length(); i++)
    {
      Character c = T.charAt(i);
      ArrayList cList = new ArrayList<>();
      if (!dp.containsKey(c))
      {
        dp.put(c, cList);
      }
      listT.add(c);
    }
    for (int j = 0; j < S.length(); j++)
    {
      Character s = S.charAt(j);
      if (T.indexOf(s) != -1)
      {
        if (!listT.contains(s) && dp.get(s).size() != 0)
        {
          dp.get(s).remove(0);
        }
        else if (listT.contains(s))
        {
          listT.remove(s);
        }
        dp.get(s).add(j);
      }
      if (listT.size() == 0)
      {
        boolean notFirst = true;

        for (List<Integer> val : dp.values())
        {
          if (notFirst)
          {
            max = val.get(val.size() - 1);
            min = val.get(0);
            notFirst = false;
          }
          if (max < val.get(val.size() - 1))
          {
            max = val.get(val.size() - 1);
          }
          if (min > val.get(0))
          {
            min = val.get(0);
          }
        }
        if ((max - min) < (end - start))
        {
          start = min;
          end = max;
        }
      }
    }

    if (listT.size() != 0)
    {
      return "";
    }
    else
    {
      return S.substring(start, end + 1);
    }
  }
}

File: Python/Minimum Window Substring.py
class Solution:
    # @return a string
    def minWindow(self, S, T):
        dp = {}
        for char in T:
            dp[char] = []
        listT = list(T)
        start = 0
        end = len(S)
        for i in range(len(S)):
            if S[i] in T:
                if S[i] not in listT and dp[S[i]] != []:
                    dp[S[i]].pop(0)
                elif S[i] in listT:
                    listT.remove(S[i])
                dp[S[i]].append(i)
            if listT == []:
                maximum = max([x[-1] for x in dp.values()])
                minimum = min([x[0] for x in dp.values()])
                if maximum-minimum+1 < end-start+1:
                    start = minimum
                    end = maximum
            print listT, dp,start, end+1
        if listT != []:
            return ""
        else:
            return S[start:end+1]
s = Solution()
print s.minWindow("AACCCBEABCODEBANC", "AB")